crates/flow-web/FIELD_SELECTION.md

```diff
@@ -0,0 +1,200 @@
+
+## Overview
+
+The `evaluate_field_selection` function provides access to Estuary's improved field selection logic from web interfaces. It evaluates which fields should be materialized based on collection projections, user configuration, connector constraints, and existing materializations.
+
+## Function Signature
+
+```typescript
+function evaluate_field_selection(input: FieldSelectionInput): FieldSelectionResult
+```
+
+## Input Format
+
+```typescript
+interface FieldSelectionInput {
+  collectionKey: string[];                    // Collection key JSON pointers (e.g., ["/id", "/timestamp"])
+  collectionProjections: Projection[];        // Available fields from the collection
+  liveSpec?: MaterializationBinding;          // Existing materialization (if updating)
+  model: MaterializationBinding;              // User's desired configuration
+  validated: ValidatedBinding;                // Connector validation results
+}
+```
+
+## Output Format
+
+```typescript
+interface FieldSelectionResult {
+  outcomes: FieldOutcome[];                   // Per-field selection details
+  selection: FieldSelection;                  // Final materialization configuration
+  hasConflicts: boolean;                      // Whether conflicts need resolution
+}
+
+interface FieldOutcome {
+  field: string;                              // Field name
+  select?: SelectOutput;                      // Structured selection reason (if selected)
+  reject?: RejectOutput;                      // Structured rejection reason (if rejected)
+  isUnsatisfiable?: boolean;                  // True when conflict has ConnectorUnsatisfiable reject reason
+}
+
+interface SelectOutput {
+  reason: SelectReason;                       // Structured selection reason
+  detail: string;                             // Human-readable description
+}
+
+interface RejectOutput {
+  reason: RejectReason;                       // Structured rejection reason
+  detail: string;                             // Human-readable description
+}
+
+interface FieldSelection {
+  keys: string[];                             // Fields used as primary key
+  values: string[];                           // Fields materialized as values
+  document: string;                           // Field storing full document (if any)
+  fieldConfigJsonMap: Record<string, string>; // Per-field connector configuration
+}
+```
+
+## Field Selection Reasons
+
+### Selection Reasons (strongest to weakest)
+- **GroupByKey**: Part of the materialization group-by key
+- **CurrentDocument**: Currently used to store the document
+- **UserRequires**: Required by user's field selection
+- **ConnectorRequires**: Required by connector
+- **PartitionKey**: Collection partition key
+- **CurrentValue**: Part of current materialization
+- **UserDefined**: User-projected field
+- **ConnectorRequiresLocation**: Location-based connector requirement
+- **CoreMetadata**: Essential system fields
+- **DesiredDepth**: Within user-specified depth
```

Comment:
Which of these selection reasons indicate that the field is _required_? Or will this function only determine whether a field is _selected_ or _excluded_? Should constraint types be leveraged to determine whether a _selected_ field is _required_?

Reply:
There are multiple reasons why a field can be "required": UserRequires -- which is the `require` portion of the model; ConnectorRequires, ConnectorRequiresLocation

This routine is giving you the final outcome of field selection: whether it will be selected or not.

I _don't_ envision the UI using the structured `reason` fields on `SelectOutput` / `RejectOutput`, at least not at first. I **am** expecting that the `outcome.select.reason` and `outcome.reject.reason` would be displayed, though, perhaps as a hover or something.